resolver/resolver.go

```diff
@@ -257,6 +257,17 @@ type Target struct {
 	URL url.URL
 }
 
+// ToEndpoint retrieves endpoint without leading "/" from either `URL.Path`
+// or `URL.Opaque`. The latter is set when the former is empty.
+func (t Target) ToEndpoint() string {
```

Comment:
Sounds good - I'll make these changes.

Why are we moving away from the receiver function approach?

Reply:
That probably should by `*Target` to avoid the copy, though.

Reply:
All our existing usages are `resolver.Target` and not `*resolver.Target`. Is avoiding a copy here worth the effort of having a `&` at all callsites?

Reply:
Oh that's true, we won't get auto-coercion like we would with a `func (*Target) GetEndpoint()`.

Option 6: what if we un-deprecate `Endpoint` and set it by this same method instead?  It seems like `Endpoint` is what all resolvers should be using anyway, as opposed to `URL.Path` or `URL.Opaque`, which are much more cumbersome to use, as all resolvers presumably need to support both.

Reply:
FYI - I'll hold off on committing additional changes until a decision is made.

---

In the interim, I raised some questions below with the intent of helping us find the best approach. I hope this does not interfere with the process!

1. > what if we un-deprecate `Endpoint`...

    Would doing so conceptually conflict with the purpose of `target.URL` since value of `Endpoint` is derived from `target.URL`?

2. > It seems like `Endpoint` is what all resolvers should be using anyway, as opposed to `URL.Path` or `URL.Opaque`, which are much more cumbersome to use, as all resolvers presumably need to support both.

    I don't understand what you mean. Aren't they equivalent since resolvers ultimately use `URL.Path` or `URL.Opaque` for endpoint (e.g., [`parseTarget` in clientconn.go](https://github.com/grpc/grpc-go/blob/0e5421c1e5331f07dff52801ff24abe9ca188b2e/clientconn.go#L1605-L1614))?

Reply:
> Would doing so conceptually conflict with the purpose of target.URL since value of Endpoint is derived from target.URL?

I don't think so.  We're doing RFC 3986 parsing of the target string into a `url.URL`, then we're also trying to provide something to free resolvers from the burden of dealing with the Go internals of `url.URL`s: namely placing the same conceptual data in either `Path` or `Opaque` depending on the URI's format.

IMO the only options I like so far are:

1. Continue to provide this `Endpoint` field so resolvers can just use it directly.
2. Provide a `func EndpointFromURL(url.URL *url) string` function.
3. Let the name resolvers deal with it.  Do this "check one; if empty check the other" behavior everywhere.

(1) is easy to use, and has little drawback, given that pretty much all resolvers need this.  (2) is harder to use than (1), but accomplishes the same thing.  The one benefit I see of (2) over (1) is that it doesn't "duplicate" any data, but that seems unimportant here.  (3) distributes the problem, but at least we don't have to do any unnecessary work.

> I don't understand what you mean. Aren't they equivalent since resolvers ultimately use URL.Path or URL.Opaque for endpoint (e.g., parseTarget in clientconn.go)?

Yes, they ultimately are still using `URL.Path`/`URL.Opaque` if they use `target.Endpoint`.  And this was intended to be an equivalent solution.

> Why are we moving away from the receiver function approach?

My concern with the method was the name (`ToEndpoint()`): we aren't _converting_ a target to an endpoint, we are just returning the endpoint portion of it.  We probably would have named it `Endpoint()`, but that's already taken by the existing field.  The next best choice would be `GetEndpoint()` but Go is quite opinionated against `Get`ters.

Reply:
I like option (2), since it moves us to a better state where the `Endpoint` field is not used. I agree that in resolver and balancer implementations, `target.Endpoint()` is a little easier to use than `resolver.EndpointFromURL(target.URL)`, but I don't see it as that big of a deal.

@dfawley : Are you fine with us proceeding with option (2).

Reply:
> I like option (2), since it moves us to a better state where the `Endpoint` field is not used.

Why do you consider this a "better state"?  Was there a problem with this `Endpoint` field, besides the fact that it used to be computed manually (i.e. not using `url.Parse` at all)?  Now that it is, I think it's not really a problem to have it.  I'm not a huge fan of it since it's duplicated data.

I'd also probably be OK with just deleting the field now, and adding a `target.Endpoint()` method.  It's a breaking change, but it's a field that's been marked as "deprecated" for some time, and in an experimental package as well.  Most likely some users will hate us for it, but we'll want to be making breaking changes here eventually.

Reply:
> Why do you consider this a "better state"?

For the reason that we can avoid duplicated data and also avoid actively using a deprecated field.

> I'd also probably be OK with just deleting the field now, and adding a target.Endpoint() method.

I would be fine with this too.

Reply:
@kylejb : Looks like we have arrived at a decision :)
Would you like to take this forward.
Thanks.

Reply:
@easwars - done :)